packages/next/src/server/require-hook.ts

```diff
@@ -117,3 +117,29 @@ mod._resolveFilename = function (
 
   // We use `bind` here to avoid referencing outside variables to create potential memory leaks.
 }.bind(null, resolveFilename, hookPropertyMap)
+
+// This is a hack to make sure that if a user requires a Next.js module that wasn't bundled
+// that needs to point to the rendering runtime version, it will point to the correct one.
+// This can happen on `pages` when a user requires a dependency that uses next/image for example.
+// This is only needed in production as in development we fallback to the external version.
+if (
+  process.env.NODE_ENV !== 'development' &&
+  process.env.__NEXT_PRIVATE_RENDER_RUNTIME &&
+  !process.env.TURBOPACK
+) {
+  const currentRuntime = `${
+    process.env.__NEXT_PRIVATE_RENDER_RUNTIME === 'pages'
+      ? 'next/dist/compiled/next-server/pages.runtime'
+      : 'next/dist/compiled/next-server/app-page.runtime'
+  }.prod`
+
+  mod.prototype.require = function (request: string) {
+    if (request.endsWith('.shared-runtime')) {
+      const base = path.basename(request, '.shared-runtime')
+      const camelized = base.replace(/-([a-z])/g, (g) => g[1].toUpperCase())
+      const instance = originalRequire.call(this, currentRuntime)
```

Comment:
Can this be cached in a map? Might not be that frequent though.

Reply:
I think it should be fine in the context of a single req